pkg/runtime/host.go

```diff
@@ -8,25 +8,37 @@ import (
 )
 
 const (
+	// HostIPEnvVar is the environment variable to override host's chosen IP address.
 	HostIPEnvVar = "DAPR_HOST_IP"
 )
 
+// GetHostAddress selects a valid outbound IP address for the host.
 func GetHostAddress() (string, error) {
 	if val, ok := os.LookupEnv(HostIPEnvVar); ok && val != "" {
 		return val, nil
 	}
 
-	addrs, err := net.InterfaceAddrs()
+	// Use udp so no handshake is made.
+	// Any IP can be used, since connection is not established, but we used a known DNS IP.
+	conn, err := net.Dial("udp", "8.8.8.8:80")
 	if err != nil {
-		return "", fmt.Errorf("error getting interface addresses: %s", err)
-	}
+		// Could not find one via a  UDP connection, so we fallback to the "old" way: try first non-loopback IPv4:
+		addrs, err := net.InterfaceAddrs()
+		if err != nil {
+			return "", fmt.Errorf("error getting interface IP addresses: %s", err)
+		}
 
-	for _, addr := range addrs {
-		if ipnet, ok := addr.(*net.IPNet); ok && !ipnet.IP.IsLoopback() {
-			if ipnet.IP.To4() != nil {
-				return ipnet.IP.String(), nil
+		for _, addr := range addrs {
+			if ipnet, ok := addr.(*net.IPNet); ok && !ipnet.IP.IsLoopback() {
+				if ipnet.IP.To4() != nil {
+					return ipnet.IP.String(), nil
+				}
 			}
 		}
+
+		return "", errors.New("could not determine host IP address")
 	}
-	return "", errors.New("could not determine non-loopback host address")
-}
\ No newline at end of file
+
+	defer conn.Close()
```

Comment:
moving `defer conn.Close()` to line 24 make sense, because I can see the returns in line28 and line34.

Reply:
It does not because you don't have guarantee of conn being set until like 42. See the error check.

Reply:
Let me know is there is any concerns with this PR.